More powerful version of ihex_byte_copy()

acolomb opened the pull request:
This one is based on and contains the changes of Pull Request #17.

This alternative version is more flexible and allows an offset to be specified, in case the user wants to copy just a part of the data. Setting the new last argument to zero behaves just as the previous implementation. A bit more error checking is added on top of that, to detect when the requested data range was not contained in the record set at all.

This one works well together with Pull Request #18 to find the most sensible values for `off` and `n` parameters.
- No unit tests yet.
- An example in the README.md would probably be nice for grokking this part of the API.
- The code to iterate through records while adjusting offset addresses, checking for premature EOF, etc. should probably be factored out and reused where applicable. I have some ideas there, but it's probably better to discuss the new parts of the API first.

acolomb committed:
Add new API function ihex_byte_copy().

The implementation is basically the same as the old ihex_mem_copy()
function.  However, the destination memory is not zeroed to avoid
overwriting content which is not specified in the ihex records.

The copy is done completely byte-wise, avoiding memory access
violations.  Record contents whose address lies beyond the given
destination range are silently discarded instead of aborting with an
IHEX_ERR_ADDRESS_OUT_OF_RANGE error code.

The new parameter off indicates the start address from which the
destination range should be filled.  Keep track of the minimum and
maximum data address within the record set and report an
IHEX_ERR_ADDRESS_OUT_OF_RANGE error if the source and destination
ranges did not overlap at all.

martin-helmich commented on include/cintelhex.h:
```diff
@@ -174,6 +176,21 @@
+/// Copy the content of a record set byte-wise.
+/** This method copies the content of a record set to a certain
+ *  location in memory.  In contrast to ihex_mem_copy(), the
+ *  destination memory is not zeroed first.  Any record contents whose
```
I'd like it better when both methods (`ihex_byte_copy` and `ihex_mem_copy`) behaved consistently. I do not insist of having both functions to implicitly zero the destination content (although changing this behaviour in `ihex_mem_copy` would be a breaking change).

martin-helmich commented on src/ihex_copy.c:
```diff
@@ -123,6 +123,80 @@ int ihex_mem_copy(ihex_recordset_t *rs, void* dst, ulong_t n,
+					"Unknown record type in record %i: 0x%02x",
+					i+1, x->ihr_type);
+		}
+	}
```
This is a lot of duplicate code; maybe there is a way to extract the common code from `ihex_mem_copy` and `ihex_byte_copy`?

acolomb replied:
As stated in the Pull Request, I wanted to agree on a public API first, then we can factor out the common code.

acolomb commented:
Rebased onto current master.

**TODO before merging:**
- Code formatting according to #32
- Add unit tests (I'm still uncertain what / how I should test)
- Bump libtool version for added API
- An example in the README.md would probably be nice for grokking this part of the API.

acolomb committed:
Fix segfault at EOF record.

Correctly check if a record was returned from ihex_rs_iterate_data()
before trying to deduce the minimum address.
